In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

### Данные взяты из открытого датасета: https://www.kaggle.com/joniarroba/noshowappointments
Набор представляет из себя записи 300000 пациентов, содержащих 15 признаков.

Необходимо создать модель, которая предсказывает вероятность того, придёт ли пациент на повторный визит или нет.

In [3]:
dataset=pd.read_csv('No-show-Issue-Comma-300k.csv')

### Для обучения модели будут использованы 11 признаков:
1. Возраст пациента;
2. Пол пациента;
3. День недели, на окторую пациент записан;
4. Есть ли диабет;
5. Есть ли алкогольная зависимость;
6. Есть ли артериальная гипертензия;
7. Есть ли когнитивные нарушения;
8. Курение;
9. Образование;
10. Есть ли туберкулез;
12. Интервал между записью и временем приёма (дней);

In [61]:
dataset

Age Gender AppointmentRegistration        ApointmentData DayOfTheWeek  \
0        19      M    2014-12-16T14:46:25Z  2015-01-14T00:00:00Z    Wednesday   
1        24      F    2015-08-18T07:01:26Z  2015-08-19T00:00:00Z    Wednesday   
2         4      F    2014-02-17T12:53:46Z  2014-02-18T00:00:00Z      Tuesday   
3         5      M    2014-07-23T17:02:11Z  2014-08-07T00:00:00Z     Thursday   
4        38      M    2015-10-21T15:20:09Z  2015-10-27T00:00:00Z      Tuesday   
5         5      F    2014-06-17T06:47:27Z  2014-07-22T00:00:00Z      Tuesday   
6        46      F    2015-06-19T19:53:58Z  2015-07-07T00:00:00Z      Tuesday   
7         4      F    2015-08-07T08:19:19Z  2015-08-21T00:00:00Z       Friday   
8        20      F    2014-10-14T11:46:36Z  2014-10-28T00:00:00Z      Tuesday   
9        51      F    2014-06-13T07:50:43Z  2014-06-17T00:00:00Z      Tuesday   
10       33      M    2015-10-01T08:43:18Z  2015-11-09T00:00:00Z       Monday   
11       58      F    2014-04-30T08:04:27Z  2014-05-22T00:00:00Z     Thursday   
12       62      F    2014-07-25T15:02:33Z  2014-08-11T00:00:00Z       Monday   
13       62      F    2014-07-08T12:56:25Z  2014-07-14T00:00:00Z       Monday   
14       38      M    2014-03-21T08:23:16Z  2014-03-27T00:00:00Z     Thursday   
15       73      M    2015-04-01T16:27:05Z  2015-04-02T00:00:00Z     Thursday   
16       48      F    2015-03-07T10:45:25Z  2015-03-30T00:00:00Z       Monday   
17       52      M    2014-06-03T10:20:37Z  2014-06-05T00:00:00Z     Thursday   
18       64      F    2014-02-17T14:32:34Z  2014-02-24T00:00:00Z       Monday   
19        1      M    2014-04-01T13:39:00Z  2014-04-02T00:00:00Z    Wednesday   
20       33      F    2014-03-12T10:26:26Z  2014-03-15T00:00:00Z     Saturday   
21       44      F    2015-04-08T12:31:28Z  2015-04-10T00:00:00Z       Friday   
22        5      F    2015-11-11T09:44:38Z  2015-11-12T00:00:00Z     Thursday   
23       69      F    2014-11-24T08:25:35Z  2014-11-28T00:00:00Z       Friday   
24       16      F    2014-10-23T16:49:42Z  2014-10-28T00:00:00Z      Tuesday   
25        8      M    2015-05-11T14:56:01Z  2015-05-25T00:00:00Z       Monday   
26        7      M    2015-04-01T08:45:19Z  2015-04-16T00:00:00Z     Thursday   
27       51      F    2014-07-29T10:33:20Z  2014-07-31T00:00:00Z     Thursday   
28       29      F    2014-08-18T10:30:16Z  2014-08-20T00:00:00Z    Wednesday   
29       59      M    2014-02-12T08:29:48Z  2014-03-14T00:00:00Z       Friday   
...     ...    ...                     ...                   ...          ...   
299970   74      F    2015-09-02T17:17:18Z  2015-09-09T00:00:00Z    Wednesday   
299971    9      M    2014-07-01T07:21:56Z  2014-07-03T00:00:00Z     Thursday   
299972   28      M    2015-07-14T09:07:11Z  2015-07-20T00:00:00Z       Monday   
299973   47      M    2014-09-03T18:43:46Z  2014-09-11T00:00:00Z     Thursday   
299974   61      M    2015-04-09T12:55:52Z  2015-04-10T00:00:00Z       Friday   
299975   13      M    2015-12-22T08:13:21Z  2015-12-23T00:00:00Z    Wednesday   
299976   52      F    2014-05-19T12:16:19Z  2014-05-21T00:00:00Z    Wednesday   
299977   46      F    2014-10-01T08:16:09Z  2014-10-10T00:00:00Z       Friday   
299978   70      F    2014-06-30T09:29:04Z  2014-07-01T00:00:00Z      Tuesday   
299979   42      M    2014-03-24T11:10:44Z  2014-03-26T00:00:00Z    Wednesday   
299980    7      F    2015-10-01T16:58:44Z  2015-10-13T00:00:00Z      Tuesday   
299981   56      M    2014-11-21T17:15:48Z  2014-11-25T00:00:00Z      Tuesday   
299982   84      F    2014-11-04T13:54:47Z  2014-11-07T00:00:00Z       Friday   
299983    4      F    2015-04-29T13:46:40Z  2015-04-30T00:00:00Z     Thursday   
299984    5      F    2015-04-28T07:28:21Z  2015-05-27T00:00:00Z    Wednesday   
299985    0      F    2015-06-16T16:01:27Z  2015-07-14T00:00:00Z      Tuesday   
299986   67      F    2014-04-08T11:00:28Z  2014-04-10T00:00:00Z     Thursday   
299987   73      F    2015-09-24T18:47:

### Предобработка данных:

In [4]:
age_drop=dataset[dataset.Age<0]
dataset_1=dataset.drop(age_drop.index, axis=0)

In [19]:
X=dataset_1.iloc[:, [0, 1,4, 6,7,8,9,10,11,12,13,14]].values
y=dataset_1.iloc[:, 5].values

In [5]:
X[0]

array([19, 'M', 0, 0, 0, 0, 0, 0, 0, 0, -29], dtype=object)

In [26]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
binarizer_1=LabelEncoder()
X[:, 1]=binarizer_1.fit_transform(X[:, 1])
y=binarizer_1.fit_transform(y)

binarizer_2=LabelEncoder()
X[:, 2]=binarizer_4.fit_transform(X[:, 2])
one_hot_day_of_week=OneHotEncoder(categorical_features=[2])
X=one_hot_day_of_week.fit_transform(X).toarray()
X=X[:, 1:]

In [29]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

/home/nbuser/anaconda3_431/lib/python3.6/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [30]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train = sc.fit_transform(X_train)
X_test=sc.fit_transform(X_test)

### Содание модели искусственной нейронной сети

In [32]:
from keras.models import Sequential
from keras.layers import Dense, PReLU, LeakyReLU
from keras.optimizers import SGD, RMSprop, Adagrad

rms1=RMSprop(lr=0.001, rho=0.9, epsilon=1e-08, decay=0.0)

classifier=Sequential()
classifier.add(Dense(units=32, kernel_initializer='uniform', input_dim=X_train.shape[1]))
classifier.add(PReLU())
classifier.add(Dense(units=16, kernel_initializer='uniform', input_dim=X_train.shape[1]))
classifier.add(PReLU())
classifier.add(Dense(units=8, kernel_initializer='uniform', input_dim=X_train.shape[1]))
classifier.add(PReLU())
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))


classifier.compile(optimizer=rms1, loss='binary_crossentropy', metrics=['accuracy'])

Using TensorFlow backend.


### Обучение модели:

In [34]:
classifier.fit(X_train, y_train, epochs=10000, batch_size=1000, verbose=2, class_weight = 'auto')

Epoch 1/10000
0s - loss: 0.5989 - acc: 0.6984
Epoch 2/10000
1s - loss: 0.5989 - acc: 0.6985
Epoch 3/10000
1s - loss: 0.5989 - acc: 0.6984
Epoch 4/10000
0s - loss: 0.5989 - acc: 0.6984
Epoch 5/10000
1s - loss: 0.5988 - acc: 0.6984
Epoch 6/10000
1s - loss: 0.5988 - acc: 0.6985
Epoch 7/10000
1s - loss: 0.5988 - acc: 0.6985
Epoch 8/10000
1s - loss: 0.5988 - acc: 0.6985
Epoch 9/10000
1s - loss: 0.5988 - acc: 0.6985
Epoch 10/10000
1s - loss: 0.5988 - acc: 0.6986
Epoch 11/10000
0s - loss: 0.5987 - acc: 0.6986
Epoch 12/10000
0s - loss: 0.5988 - acc: 0.6985
Epoch 13/10000
1s - loss: 0.5987 - acc: 0.6985
Epoch 14/10000
1s - loss: 0.5987 - acc: 0.6986
Epoch 15/10000
1s - loss: 0.5987 - acc: 0.6985
Epoch 16/10000
0s - loss: 0.5987 - acc: 0.6986
Epoch 17/10000
0s - loss: 0.5987 - acc: 0.6985
Epoch 18/10000
1s - loss: 0.5987 - acc: 0.6985
Epoch 19/10000
1s - loss: 0.5987 - acc: 0.6986
Epoch 20/10000
1s - loss: 0.5986 - acc: 0.6987
Epoch 21/10000
1s - loss: 0.5986 - acc: 0.6987
Epoch 22/10000
1s - lo

KeyboardInterrupt: 

### Валидация модели на тестовой выборке (на данных, которые модель "не видела")

In [82]:
y_pred=classifier.predict_classes(X_test, verbose=2)

### точность модели = количество правильных прогнозов / общее количество прогнозов

In [83]:
accuracy=classifier.evaluate(X_test, y_test, batch_size=1, verbose=2)
print(' Точность модели: %.1f'% (accuracy[1]*100), "%")

 Точность модели: 69.6 %


### Точность модели составила: 69.6 %

### Для данного набора данных самая высокая точность модели на тестовой выборке составила 71% (из опубликованных результатов на сайте kaggle). Это можно объяснить: 
### 1. Искажением некоторых признаков: в переменной возраст встречались отрицательные и нулевые значения;
### 2. Нерелеватными признаками: теоретически, для обучения модели логичнее взять такие признаки, как: 
    - общее количество консультаций, 
    - количество бонусных баллов, 
    - общее количество отменённых визитов к врачу, 
    - длительность обслуживания пациента, 
    - количество госпитализаций и т.п. 
##### Релеватные признаки позволят повысить предиктивную точность модели.